alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,263 @@
+            name: the algorithm name, "MISCAlgorithm"
+        """
+
+        super(MISCAlgorithm, self).__init__(
+            train_state_spec=feature_spec, name=name)
+
+        feature_dim = tf.nest.flatten(feature_spec)[0].shape[-1]
+
+        if misc_layerx is None:
+            misc_layerx = tf.keras.layers.Dense(
+                hidden_size, input_shape=soc_spec.shape, activation=None)
+            misc_layery = tf.keras.layers.Dense(
+                hidden_size, input_shape=soi_spec.shape, activation=None)
+            misc_layero = tf.keras.layers.Dense(
+                1, input_shape=(None, None, hidden_size), activation=None)
+
+        self._misc_layerx = misc_layerx
+        self._misc_layery = misc_layery
+        self._misc_layero = misc_layero
+
+        self._traj_spec = traj_spec
+        self._buffer_size = buffer_size
+        self._buffer = DataBuffer(self._traj_spec, capacity=self._buffer_size)
+        self._mi_r_scale = mi_r_scale
+        self._n_objects = n_objects
+        self._empowerment = empowerment
+        self._env_name = env_name
+
+    def split_observation_tf(self, o):
+
+        dimo = o.get_shape().as_list()[-1]
+
+        if self._env_name in ['SocialBot-PlayGround-v0']:
+            if dimo == 21:
+                task_specific_ob, agent_pose, agent_vel, internal_states, action = tf.split(
+                    o, [3, 6, 6, 4, 2], axis=-1)
+
+                agent_pose_1, agent_pose_2 = tf.split(
+                    agent_pose, [3, 3], axis=-1)
+                joint_pose_1, joint_pose_2, joint_vel_1, joint_vel_2 = tf.split(
+                    internal_states, [1, 1, 1, 1], axis=-1)
+                joint_1 = tf.concat([joint_pose_1, joint_vel_1], axis=-1)
+                joint_2 = tf.concat([joint_pose_2, joint_vel_2], axis=-1)
+
+                if self._n_objects == 1:
+                    obs_achieved_goal = task_specific_ob
+                    obs_excludes_goal = agent_pose_1
+                    if self._empowerment:
+                        obs_excludes_goal = action
+                elif self._n_objects == 0:
+                    obs_achieved_goal = joint_1
+                    obs_excludes_goal = joint_2
+
+                return (obs_excludes_goal, obs_achieved_goal)
+
+            elif (dimo == 24) and (self._n_objects == 2):
+                task_specific_ob_1, task_specific_ob_2, agent_pose, agent_vel, internal_states, action = tf.split(
+                    o, [3, 3, 6, 6, 4, 2], axis=-1)
+
+                agent_pose_1, agent_pose_2 = tf.split(
+                    agent_pose, [3, 3], axis=-1)
+
+                obs_achieved_goal_1 = task_specific_ob_1
+                obs_achieved_goal_2 = task_specific_ob_2
+                obs_excludes_goal = agent_pose_1
+
+                return (obs_excludes_goal, obs_achieved_goal_1,
+                        obs_achieved_goal_2)
+            else:
+                print(
```

Comment:
use `raise` instead so that people can know where the error is from by looking at the stack trace.

Reply:
use as required